tm2/pkg/crypto/keys/client/verify.go

```diff
@@ -39,63 +146,227 @@ func (c *VerifyCfg) RegisterFlags(fs *flag.FlagSet) {
+	}
+
+	// get the transaction to verify.
+	tx, err := getTransaction(cfg, io)
 	if err != nil {
 		return err
 	}
+
+	// verify signature.
+	sig, err := getSignature(cfg, tx)
+	if err != nil {
+		return err
+	}
+
+	// Get account number and sequence if needed.
+	getTxParameters(ctx, cfg, info, io)
+
+	// Get the bytes to verify
+	signBytes, err := tx.GetSignBytes(
+		cfg.ChainID,
+		cfg.AccountNumber,
+		cfg.AccountSequence,
+	)
+	if err != nil {
+		return fmt.Errorf("unable to get signature bytes, %w", err)
+	}
+
+	err = kb.Verify(info.GetName(), signBytes, sig)
+	if err == nil {
+		if !cfg.RootCfg.BaseOptions.Quiet {
+			io.Println("Valid signature!")
+		}
+	}
+	return err
+}
+
+func getTransaction(cfg *VerifyCfg, io commands.IO) (*std.Tx, error) {
 	var msg []byte
 
 	// read document to sign
-	if docpath == "" { // from stdin.
+	if cfg.DocPath == "" { // from stdin.
 		msgstr, err := io.GetString(
 			"Enter document to sign.",
 		)
 		if err != nil {
-			return err
+			return nil, err
 		}
 		msg = []byte(msgstr)
 	} else { // from file
-		msg, err = os.ReadFile(docpath)
+		var err error
+		msg, err = os.ReadFile(cfg.DocPath)
 		if err != nil {
-			return err
+			return nil, err
 		}
 	}
 
-	// validate document to sign.
-	// XXX
+	// Unmarshal transaction Amino JSON
+	var tx std.Tx
+	if err := amino.UnmarshalJSON(msg, &tx); err != nil {
+		return nil, fmt.Errorf("unable to unmarshal transaction, %w", err)
+	}
+	return &tx, nil
+}
 
-	// verify signature.
-	err = kb.Verify(name, msg, sig)
-	if err == nil {
-		io.Println("Valid signature!")
+func getSignature(cfg *VerifyCfg, tx *std.Tx) ([]byte, error) {
+	// Exclude -sigpath and -signature flags set at the same time.
+	if cfg.SigPath != "" && cfg.Signature != "" {
+		return nil, errors.New("only one of -sigpath or -signature flags can be set")
+	}
+
+	// from -sigpath flag
+	if cfg.SigPath != "" {
+		sigbz, err := os.ReadFile(cfg.SigPath)
+		if err != nil {
+			return nil, err
+		}
+
+		// Unmarshal transaction Amino JSON
+		var sig std.Signature
+		if err := amino.UnmarshalJSON(sigbz, &sig); err != nil {
+			return nil, fmt.Errorf("unable to unmarshal signature, %w", err)
+		}
+
+		if sig.Signature == nil {
+			return nil, errors.New("no signature found in the signature file")
+		}
+
+		return sig.Signature, nil
+	}
+
+	// from -signature flag
+	if cfg.Signature != "" {
+		sig, err := base64.StdEncoding.DecodeString(cfg.Signature)
+		if err != nil {
+			return nil, fmt.Errorf("unable to decode signature, %w", err)
+		}
+		return sig, nil
+	}
+
+	// default: from tx
+	if len(tx.Signatures) > 0 {
+		return tx.Signatures[0].Signature, nil
+	}
+
+	return nil, errors.New("no signature found in the transaction")
+}
+
+func getTxParameters(ctx context.Context, cfg *VerifyCfg, info keys.Info, io commands.IO) {
+	// Query the chain if -offline=false and account number or sequence are equal to 0.
+	if !cfg.Offline && (cfg.AccountNumber == 0 || cfg.AccountSequence == 0) {
```

Comment:
Please use early return.
```suggestion
	if cfg.Offline || (cfg.AccountNumber != 0 && cfg.AccountSequence != 0) {
		return
	}
	// ...
```